src/check_bird_protocols.py

```diff
@@ -0,0 +1,221 @@
+
+    if result[ProtocolStates.down]:
+        return (
+            ExitCodes.critical,
+            'Some protocols are down: {}'
+            .format(', '.join(result[ProtocolStates.down]))
+        )
+    elif result[ProtocolStates.disabled]:
+        return (
+            ExitCodes.warning,
+            'Some protocols are disabled: {}'
+            .format(', '.join(result[ProtocolStates.disabled]))
+        )
+    elif result[ProtocolStates.unknown]:
+        return (
+            ExitCodes.unknown,
+            'Unknown protocols are seen: {}'
+            .format(', '.join(result[ProtocolStates.unknown]))
+        )
+    else:
+        return (ExitCodes.ok, 'All protocols are up')
+
+
+def check_bgp_protocol(protocol):
+    def is_bgp_up(protocol):
+        is_established = protocol['info'] == 'Established'
+        is_up = protocol['state'] == 'up'
+        return is_up and is_established
+
+    def is_bgp_disabled(protocol):
+        state_down = protocol['state'] == 'down'
+        return state_down
+
+    if is_bgp_up(protocol):
+        return ProtocolStates.up
+    elif is_bgp_disabled(protocol):
+        return ProtocolStates.disabled
+    else:
+        return ProtocolStates.down
+
+
+def check_ospf_protocol(protocol):
+    def is_ospf_up(protocol):
+        is_up = protocol['state'] == 'up'
+        is_running = protocol['info'] == 'Running'
+        return is_up and is_running
+
+    def is_ospf_disabled(protocol):
+        is_down = protocol['state'] == 'down'
+        return is_down
+
+    if is_ospf_up(protocol):
+        return ProtocolStates.up
+    elif is_ospf_disabled(protocol):
+        return ProtocolStates.disabled
+    else:
+        return ProtocolStates.down
+
+
+def check_other_protocol(protocol):
+    if protocol['state'] == 'up':
+        return ProtocolStates.up
+    elif protocol['state'] == 'down':
+        return ProtocolStates.disabled
+    else:
+        return ProtocolStates.down
+
+
+def parse_birdc_output(output):
+    # Matching protocol lines from BIRD:
+    # BIRD 1.6.3 ready.
+    # name     proto    table    state  since       info
+    # direct1  Direct   master   up     2020-02-27
+    # b_af     BGP      master   up     2020-02-28  Established
+    # b_lb     BGP      master   start  18:43:46    Active        Socket: conn
+    # b_home   BGP      master   down   2020-02-29
+    # We are extracting the protocol name, state, date/time since last
+    # change and the additional information about the state of the session.
+    lines = output.splitlines()
+    lines = lines[2:]
+
+    tokenized_lines = [
+        [field.strip() for field in line.split(None, 5)]
+        for line in lines
+    ]
+    protocols_dicts = [
+        build_dict(line) for line in tokenized_lines
+    ]
+
+    return protocols_dicts
+
+
+def build_dict(line):
+    # The next line ensures that we always have 6 elements in the line,
+    # as the tokenization of protocols that are disabled yeld only 5 elements.
```

Comment:
```suggestion
    # as the tokenization of protocols that are disabled yield only 5 elements.
```
of
```suggestion
    # as the tokenization of protocols that are disabled hold only 5 elements.
```
Not sure if you wanted to say `yield` or `hold`. ^^

Reply:
I meant yield, but my keyboard ate the I :joy: